MetaMask/test-dapp
# feat: add Wallet Connect  support

## Context
Needed for testing the PPOM module in Mobile + Wallet Connect

## What
Adds wallet connect support to the test dapp.

## How
- A button triggers the Wallet Connect Modal. Using the Wallet Connect option, you can scan the QR code using your phone based MetaMask Wallet, to start getting transaction/signatures requests triggered from the test dapp on your phone wallet.
- Whenever the WalletConnect connection is established, a new provider is added. You can then choose to use MetaMask or WalletConnect providers alternatively to trigger your Extension or your Mobile wallet
- For Wallet Connect, we only have a subset of functionalities enabled for now: Sends/Signatures but excluding Contract Deployments, as needs some more investigation on how to get the contract address. This can be extended in the future (not currently needed for the purposes of PPOM).
- When we use the `?contract` URI param on the URL, all the buttons (except deploy contracts) are also enabled for Wallet Connect to use, in the same was as it's done for Extension

## Screenshots


https://github.com/MetaMask/test-dapp/assets/54408225/57fda92e-6e76-40b5-9f7c-48c81ee03db3


![Screenshot from 2023-12-11 16-19-41](https://github.com/MetaMask/test-dapp/assets/54408225/17d74652-e5d0-4866-8d94-b0d01b96bd57)



## Future Work
- Refactor button list and in general the index.js file, as it starts to grow big. We could better organize the different dapp functionalities in the future
- Add Deploy Smart Contract functionality for Wallet Connect
- Add SDK support in the same way -- this will enable to test PPOM with SDK

## Manual Testing Steps
1. Build the test dapp locally `yarn` and `yarn start`
2. Go to the browser

### Testing the Connection
3. Click Wallet Connect and select Wallet Connect option
4. With your phone, scan the QR code within MM wallet
    - See connection prompts to the phone
5. Accept connect
    - See Wallet Connect Provider is added on top of the test-dapp
    - See current selected provider is now Wallet Connect
    - See Wallet Connect buttons are enabled (Sends and signatures)

### Testing tx/signatures with WC
6. Trigger any tx/signature
    - See the tx/signature is triggered on your phone

### Testing Network switch handling
7. Switch networks
    -  See status network on the test dapp is updated

### Testing switching providers back and forth (Extension-Mobile)
8. Click Use Metamask Main
    - See current selected provider is now MetaMask Main
    - See Extension Connected buttons are all enabled
9. Trigger another tx/signature
    - See now the Extension wallet is triggered
10. Click Use Wallet Connect
    - See current selected provider is back to Wallet Connect

### Testing the old functionality
Furthermore, you can check that the old functionality is 100% preserved when using MetaMask Extension provider.
